File: src/main/java/platform/ecommerce/service/product/ProductService.java
package platform.ecommerce.service.product;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import platform.ecommerce.domain.product.Product;
import platform.ecommerce.domain.product.ProductImage;
import platform.ecommerce.domain.product.ProductOption;
import platform.ecommerce.dto.request.product.*;

/**
 * Product domain service interface.
 * Returns entities for ApplicationService to convert to DTOs.
 */
public interface ProductService {

    /**
     * Create a new product.
     * @return created Product entity
     */
    Product createProduct(Long sellerId, ProductCreateRequest request);

    /**
     * Get product by ID.
     * @return Product entity
     */
    Product getProduct(Long productId);

    /**
     * Search products with conditions.
     * @return page of Product entities
     */
    Page<Product> searchProducts(ProductSearchCondition condition, Pageable pageable);

    /**
     * Update product details.
     * @return updated Product entity
     */
    Product updateProduct(Long productId, ProductUpdateRequest request);

    /**
     * Publish product (DRAFT -> ACTIVE).
     * @return updated Product entity
     */
    Product publishProduct(Long productId);

    /**
     * Discontinue product.
     */
    void discontinueProduct(Long productId);

    /**
     * Delete product (soft delete).
     */
    void deleteProduct(Long productId);

    /**
     * Add option to product.
     * @return created ProductOption entity
     */
    ProductOption addOption(Long productId, ProductOptionRequest request);

    /**
     * Remove option from product.
     */
    void removeOption(Long productId, Long optionId);

    /**
     * Update option stock.
     * @return updated ProductOption entity
     */
    ProductOption updateOptionStock(Long productId, Long optionId, int stock);

    /**
     * Add image to product.
     * @return created ProductImage entity
     */
    ProductImage addImage(Long productId, String imageUrl, String altText);

    /**
     * Remove image from product.
     */
    void removeImage(Long productId, Long imageId);

    /**
     * Decrease stock (for order).
     */
    void decreaseStock(Long productId, Long optionId, int quantity);

    /**
     * Increase stock (for cancellation).
     */
    void increaseStock(Long productId, Long optionId, int quantity);
}

File: src/main/kotlin/platform/ecommerce/service/ProductService.kt
package platform.ecommerce.service

import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import platform.ecommerce.domain.Product
import platform.ecommerce.domain.ProductOption
import platform.ecommerce.dto.request.ProductCreateRequest
import platform.ecommerce.dto.request.ProductOptionRequest
import platform.ecommerce.exception.ProductNotFoundException
import platform.ecommerce.repository.ProductRepository
import platform.ecommerce.utils.Logger.Companion.logger

@Service
class ProductService(
    private val productRepository: ProductRepository
) {

    @Transactional
    fun createProduct(request: ProductCreateRequest): Product {
        val product = Product.create(
            sku = request.sku,
            name = request.name,
            description = request.description,
            price = request.price,
            imageUrl = request.imageUrl
        )

        val savedProduct = productRepository.save(product)

        logger.info { "Product created: productId=${savedProduct.id}, sku=${savedProduct.sku}" }

        return savedProduct
    }

    @Transactional
    fun deleteProduct(productId: Long) {
        val product = productRepository.findById(productId)
            ?: throw ProductNotFoundException("Product not found: productId=$productId")

        product.softDelete()
    }

    @Transactional
    fun addProductOption(productId: Long, request: ProductOptionRequest): ProductOption {
        val product = productRepository.findById(productId)
            ?: throw ProductNotFoundException("Product not found: productId=$productId")

        product.addOption(
            optionName = request.optionName,
            stockQuantity = request.stockQuantity
        )

        logger.info { "Option added: productId=$productId, optionName=${request.optionName}" }

        return product.options.last()
    }

    @Transactional
    fun removeProductOption(productId: Long, optionId: Long) {
        val product = productRepository.findById(productId)
            ?: throw ProductNotFoundException("Product not found")

        product.removeOption(optionId)

        logger.info{ "Option removed: productId=$productId, optionId=$optionId" }
    }

    @Transactional(readOnly = true)
    fun getProduct(productId: Long): Product {
        return productRepository.findWithOptionsById(productId)
            ?: throw ProductNotFoundException("Product not found")
    }

    @Transactional(readOnly = true)
    fun getAllProducts(): List<Product> {
        return productRepository.findAll()
    }
}
